### PR TITLE
Implement LoRA for efficient finetuning

### PR DESCRIPTION
**Pull request by rasbt**
Implements LoRA for efficient finetuning of parrot models

- [x] add finetuning script
- [x] add the howto guide 
- [x] add tests
- [x] and generate script.
- [ ] ~~Support Falcon (MQA, GQA)~~: left for a separate PR
- [x] add `ds_config` for deepspeed

### CODE REVIEW

**Comment by rasbt** [comment]
Just ran a full Alpaca finetuning round for StableLM 3B and the results looks great. Loss converges to ~1 and it generates very sensible outputs. 

It takes 45 min on a A100 for Alpaca. 

Should be ready to review (and hopefully merge) when you have time @lantiga @awaelchli @carmocca

File: finetune/lora.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+    checkpoint_path = checkpoint_dir / "lit_model.pth"
+   
+    fabric.print(f"Loading model {str(checkpoint_path)!r} with {config.__dict__}")
+    
+    #with fabric.init_module():
+    #    model = Parrot(config)
+    #with lazy_load(checkpoint_dir / "lit_model.pth") as checkpoint:
+    #    # strict=False because missing keys due to adapter weights not contained in state dict
+    #    model.load_state_dict(checkpoint, strict=False)
```
Comment (carmocca) [inline finetune/lora.py:80]: What is the issue with `lazy_load`?

File: finetune/lora.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+
+def main(
+    data_dir: Path = Path("data/alpaca"),
+    checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
+    out_dir: Path = Path("out/lora/alpaca"),
+    precision: Literal["bf16-true", "32-true", "bf16-mixed"] = "bf16-true",
+):
+
+    print(checkpoint_dir)
```
Comment (carmocca) [inline finetune/lora.py:57]: Lefover print? The "loading model" print will include this alreay

File: finetune/lora.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+# Hyperparameters
+learning_rate = 3e-4
+batch_size = 128
+micro_batch_size = 4
+gradient_accumulation_iters = batch_size // micro_batch_size
+assert gradient_accumulation_iters > 0
+max_iters = 50000 * 3 // micro_batch_size
+weight_decay = 0.0
+max_seq_length = 256  # see scripts/prepare_alpaca.py
```
Comment (carmocca) [inline finetune/lora.py:43]: Can you include the changes from #127?

File: finetune/lora.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        if step_count <= warmup_iters:
+            # linear warmup
+            lr = learning_rate * step_count / warmup_iters
+            for param_group in optimizer.param_groups:
+                param_group['lr'] = lr
+
+        t0 = time.time()
+
+        input_ids, targets = get_batch(fabric, train_data)
+        with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_iters != 0)):
```
Comment (carmocca) [inline finetune/lora.py:130]: Can you incorporate the changes from #110?

File: finetune/lora.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    fabric.print(f"Number of trainable parameters: {num_params}")
+
+    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
+    model, optimizer = fabric.setup(model, optimizer)
+    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir, out_dir)
+
+    # Save the final LoRA checkpoint at the end of training
+    checkpoint = lora_state_dict(model)
+
+    fabric.save(os.path.join(out_dir, "lit-model-lora-finetuned.pth"), checkpoint)
```
Comment (carmocca) [inline finetune/lora.py:98]: This could be an utility save function defined in this file, that does print, gets the lora state dict and saves it

File: finetune/lora.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+                val_loss = validate(fabric, model, val_data, tokenizer)
+                fabric.print(f"step {iter_num}: val loss {val_loss:.4f}")
+                fabric.barrier()
+
+            if step_count % save_interval == 0:
+                print(f"Saving LoRA weights to {out_dir}")
+                # We are only saving the LoRA weights
+                # TODO: Provide a function/script to merge the LoRA weights with pretrained weights
+                checkpoint = lora_state_dict(model)
+                fabric.save(os.path.join(out_dir, f"iter-{iter_num:06d}-ckpt.pth"), checkpoint)
```
Comment (carmocca) [inline finetune/lora.py:150]: This isn't how you would use `fabric.save`. See the docs: https://lightning.ai/docs/fabric/latest/api/fabric_methods.html#save

You probably want to use `torch.save` instead for now. We'll be working on a way to filter out what gets saved soon. Then you'll be able to use `fabric.save`

File: generate/lora.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    fabric = L.Fabric(devices=devices, precision=precision, strategy=strategy)
+    fabric.launch()
+
+    check_valid_checkpoint_dir(checkpoint_dir)
+
+    with open(checkpoint_dir / "lit_config.json") as fp:
+        config = Config(**json.load(fp))
+
+    if quantize is not None:
+        raise NotImplementedError("Quantization in LoRA is not supported yet")
```
Comment (carmocca) [inline generate/lora.py:77]: In this case, I would remove the argument entirely

File: generate/lora.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        if not (checkpoint_dir / model_file).is_file():
+            raise ValueError("Please run `python quantize/gptq.py` first")
+    else:
+        model_file = "lit_model.pth"
+    """
+    checkpoint_path = checkpoint_dir / model_file
+
+    fabric.print(f"Loading model {str(checkpoint_path)!r} with {config.__dict__}", file=sys.stderr)
+    t0 = time.time()
+    with fabric.init_module(), EmptyInitOnDevice(quantization_mode=quantize):
```
Comment (carmocca) [inline generate/lora.py:95]: This needs to be updated according to #131

File: generate/lora.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        model = Parrot(config)
+    fabric.print(f"Time to instantiate model: {time.time() - t0:.02f} seconds.", file=sys.stderr)
+
+    t0 = time.time()
+
+    with lazy_load(checkpoint_path) as pretrained_checkpoint, lazy_load(lora_path) as lora_checkpoint:
```
Comment (carmocca) [inline generate/lora.py:101]: Lazy loading should be done after the model is loaded, as in the other generation scripts

File: howto/finetune_lora.md
Comment (carmocca) [inline howto/finetune_lora.md:1]: Maybe update the README too?

File: lit_parrot/lora.py [hunk 10]
```diff
@@ -0,0 +1,477 @@
...
+
+
+@dataclass
+class LoRAConfig:
+    r: float = 0.0
+    alpha: float = 1.0
+    dropout: float = 0.0
+
+
+class CausalSelfAttention(parrot.CausalSelfAttention):
```
Comment (carmocca) [inline lit_parrot/lora.py:403]: Maybe not in this PR, but we should implement a mechanism that allows replacing other layers more easily instead of having to write this.

For instance, we would need this to address https://github.com/Lightning-AI/lit-llama/issues/350. And other frameworks for LoRA support this.

**Review by carmocca — COMMENTED** [review]
Thanks Seb! Good job.

Would you mind checking the failing tess?

Several comments. I can help with this once I finish some of my current tasks

File: finetune/lora.py (hunk 1 above)
Comment (rasbt) [reply finetune/lora.py:80]: @carmocca It resulted in shape issues. It seems like it couldn't load the tensors properly. I can try again an print the exact error message.

**Comment by rasbt** [comment]
Thanks for the review, I will try to address these cases tonight / tomorrow morning. Btw a question regarding the CI. It looks like it got automatically canceled, or is there an issue with my PR?
Screenshot 2023-06-11 at 11 57 36 AM

**Comment by carmocca** [comment]
Uh that's strange. Try pushing new commits and I'll debug it if it keeps happening

File: finetune/lora.py (hunk 2 above)
Comment (rasbt) [reply finetune/lora.py:57]: Yes, thanks!

**Comment by rasbt** [comment]
> Uh that's strange. Try pushing new commits and I'll debug it if it keeps happening

Seems to work now, no worries.

File: generate/lora.py (hunk 7 above)
Comment (rasbt) [reply generate/lora.py:77]: It might actually work, it's unclear. Can't test it at the moment due to import issues with bitsandbytes. Probably related to https://github.com/openai/triton/issues/1735

File: generate/lora.py (hunk 8 above)
Comment (rasbt) [reply generate/lora.py]: done

File: lit_parrot/lora.py (hunk 10 above)
Comment (rasbt) [reply lit_parrot/lora.py:403]: Future problem 😅

File: finetune/lora.py (hunk 4 above)
Comment (rasbt) [reply finetune/lora.py:148]: done

File: finetune/lora.py (hunk 3 above)
Comment (rasbt) [reply finetune/lora.py:44]: done

**Comment by rasbt** [comment]
Implemented all the suggestions @carmocca . Should be good to review.

**Comment by rasbt** [comment]
Arg, it all works fine with StableLM. But I just noticed that this causes issues with Falcon. 

```
        size mismatch for transformer.h.20.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.21.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.22.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.23.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.24.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.25.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.26.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.27.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.28.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.29.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.30.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
        size mismatch for transformer.h.31.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([13632, 4544]).
```

I think that's because of the multiquery attention. Any ideas for how to fix this?

File: README.md [hunk 11]
```diff
@@ -126,13 +126,26 @@ python scripts/prepare_alpaca.py
...
+
 [Adapter](https://arxiv.org/abs/2303.16199):
 
 ```bash
 python finetune/adapter.py
 ```
 
+or 
+
+[LoRA](https://arxiv.org/abs/2106.09685):
```
Comment (carmocca) [inline README.md:139]: Reducing vertical space

```suggestion
or [LoRA](https://arxiv.org/abs/2106.09685):
```

File: README.md [hunk 12]
```diff
@@ -126,13 +126,26 @@ python scripts/prepare_alpaca.py
...
+
+
+```bash
+python finetune/lora.py
+```
+
+
```
Comment (carmocca) [inline README.md:146]: Have you checked that this is also the case for LoRA?

File: README.md [hunk 13]
```diff
@@ -126,13 +126,26 @@ python scripts/prepare_alpaca.py
 
 2. Run the finetuning script
 
+For example, you can either use
+
 [Adapter](https://arxiv.org/abs/2303.16199):
```
Comment (carmocca) [inline README.md:131]: Let's add adapter v2 too?

**Comment by rasbt** [comment]
I just noticed this also needs the `ds_config` for deepspeed. Will add this to the PR shortly

File: howto/finetune_lora.md [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+1. Follow the instructions in the [README](../README.md) to install the dependencies.
+2. Download and convert the weights and save them in the `./checkpoints` folder as described [here](download_weights.md).
+3. Download the data and generate the instruction tuning dataset:
+
+   ```bash
+   python scripts/prepare_alpaca.py
+   ```
+
+See also: [Finetuning on an unstructured dataset](unstructured_dataset.md)
```
Comment (carmocca) [inline howto/finetune_lora.md:18]: This file doesn't exist in this repo

File: howto/finetune_lora.md [hunk 15]
```diff
@@ -0,0 +1,88 @@
+# Finetuning with LoRA
+
+[Low-rank adaption (LoRA)](https://arxiv.org/abs/2106.09685) is a technique to approximate the update to the linear layers in a LLM with a low-rank matrix factorization. This significantly reduces the number of trainable parameters and speeds up training with little impact on the final performance of the model.
+We demonstrate this method by instruction-finetuning Lit-Parrot StableLM 3B on the [Alpaca](https://github.com/tatsu-lab/stanford_alpaca) dataset on a **single RTX 3090 (24GB) GPU**.
```
Comment (carmocca) [inline howto/finetune_lora.md:4]: Did you verify these memory requirements for stableLM 3b? Seems too much

File: howto/finetune_lora.md [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+```bash
+python generate/lora.py --prompt "Recommend a movie to watch on the weekend."
+```
+Output:
+```
+I would recommend the movie The Martian (2015). It is a sci-fi movie starring Matt Damon that follows the story of...
+```
+
+If your GPU supports `bfloat16`, you can additionally pass `--dtype bfloat16` to bring the memory consumption down to ~14 GB.
```
Comment (carmocca) [inline howto/finetune_lora.md:44]: `dtype` doesn't exist, it's `precision` here (passed to fabric)

Same question about the memory requirements!

**Review by carmocca — COMMENTED** [review]
@rasbt I just saw your last comment. I have some code fixes ready to push. Do you want to push your latest changes first?

**Comment by rasbt** [comment]
Should we also change this to FSDP before merging @carmocca  or figure it later?

File: howto/finetune_lora.md (hunk 16 above)
Comment (rasbt) [reply howto/finetune_lora.md:44]: Just checked and it's actually 11 GB only for StableLM 3B :)

File: howto/finetune_lora.md (hunk 14 above)
Comment (rasbt) [reply howto/finetune_lora.md:18]: Good catch, I linked Aniket's blog post for now.

**Comment by rasbt** [comment]
Besides FSDP and Falcon, everything should be addressed now. Thanks for the thorough review!

**Review by carmocca — APPROVED** [review]
Awesome! Thanks @rasbt
